File: apps/web/src/lib/api-client/qc.ts
import { apiRequestJson, apiRequestNoContent, pathSegment } from "./client";
import type { TicketState } from "./tickets";

type QcDraft = {
  id: string;
  versionNumber: number;
  title: string;
  summary: string;
  productType: string;
  content: string;
  createdByUserId: string;
  createdAt: string;
  assets: {
    id: string;
    name: string;
    assetType: string;
    mimeType: string;
    sizeBytes: number;
    sha256: string;
    detectedMimeType: string;
    previewKind: "pdf" | "image" | "text" | "metadata";
    processingStatus: string;
    previewAvailable: boolean;
  }[];
  description: string;
  manifestHash: string;
};

export type QcProduct = {
  ticketId: string;
  reference: string;
  requesterUserId: string;
  state: TicketState;
  title: string;
  operationalQuestion: string | null;
  areaOrRegion: string | null;
  priority: string | null;
  requiredOutputFormat: string | null;
  checklistKeys: string[];
  latestDraft: QcDraft | null;
  managerNotes: string[];
  decisions: {
    id: string;
    status: "approved" | "rejected";
    reason: string;
    reviewerUserId: string;
    checklist: { key: string; passed: boolean }[];
    createdAt: string;
  }[];
  agentPreflight: {
    id: string;
    draftVersionId: string;
    status: "passed" | "blocked";
    checks: { key: string; passed: boolean; detail: string }[];
    blockers: string[];
    policyVersion: string;
    createdAt: string;
    findings: {
      id: string;
      category: string;
      severity: string;
      originalText: string;
      suggestedText: string;
      location: string;
      detail: string;
      confidence: number;
      blocking: boolean;
    }[];
  } | null;
  indexRecords: { id: string; productId: string; status: string; summary: string }[];
  disseminations: { id: string; productId: string; recipientUserId: string }[];
  feedbackRequests: { id: string; productId: string; requesterUserId: string; status: string }[];
  ingestedProduct: {
    id: string;
    reference: string;
    title: string;
    status: string;
    acgIds: string[];
  } | null;
};

export type QcQueue = {
  products: QcProduct[];
  items: QcQueueItem[];
};

export type QcQueueItem = {
  ticketId: string;
  reference: string;
  state: TicketState;
  claimStatus: "available" | "claimed_by_you" | "claimed";
};

export type QcApprovalInput = {
  checklist: Record<string, boolean>;
  classificationLevel: number;
  releasability: string[];
  handlingCaveats: string[];
  acgIds: string[];
  reason: string;
};

export async function listQcQueue(): Promise<QcQueue> {
  return apiRequestJson<QcQueue>("/api/v1/qc/queue", { method: "GET" });
}

export async function getQcProduct(ticketId: string): Promise<QcProduct> {
  return apiRequestJson<QcProduct>(`/api/v1/qc/products/${pathSegment(ticketId)}`, {
    method: "GET",
  });
}

export async function claimQcProduct(ticketId: string, csrfToken: string): Promise<QcProduct> {
  return apiRequestJson<QcProduct>(`/api/v1/qc/products/${pathSegment(ticketId)}/claim`, {
    headers: { "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function releaseQcClaim(ticketId: string, csrfToken: string): Promise<void> {
  await apiRequestNoContent(`/api/v1/qc/products/${pathSegment(ticketId)}/claim`, {
    headers: { "X-CSRF-Token": csrfToken },
    method: "DELETE",
  });
}

export async function approveQcProduct(
  ticketId: string,
  payload: QcApprovalInput,
  csrfToken: string,
): Promise<QcProduct> {
  return apiRequestJson<QcProduct>(`/api/v1/qc/products/${pathSegment(ticketId)}/approve`, {
    body: JSON.stringify(payload),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function rejectQcProduct(
  ticketId: string,
  reason: string,
  csrfToken: string,
): Promise<QcProduct> {
  return apiRequestJson<QcProduct>(`/api/v1/qc/products/${pathSegment(ticketId)}/reject`, {
    body: JSON.stringify({ reason }),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

File: apps/api/src/coeus/schemas/qc.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from coeus.schemas.routing import PriorityAssessmentResponse


class QcApprovalRequest(BaseModel):
    checklist: dict[str, bool]
    classification_level: int = Field(ge=0, le=5, validation_alias="classificationLevel")
    releasability: list[str] = Field(min_length=1, max_length=12)
    handling_caveats: list[str] = Field(
        min_length=1, max_length=12, validation_alias="handlingCaveats"
    )
    acg_ids: list[UUID] = Field(min_length=1, max_length=12, validation_alias="acgIds")
    reason: str = Field(min_length=3, max_length=1_000)


class QcRejectRequest(BaseModel):
    reason: str = Field(min_length=3, max_length=1_000)


class QcChecklistItemResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    key: str
    passed: bool


class QcDraftAssetResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    asset_id: UUID = Field(serialization_alias="id")
    name: str
    asset_type: str = Field(serialization_alias="assetType")
    mime_type: str = Field(serialization_alias="mimeType")
    size_bytes: int = Field(serialization_alias="sizeBytes")
    sha256: str
    detected_mime_type: str = Field(serialization_alias="detectedMimeType")
    preview_kind: str = Field(serialization_alias="previewKind")
    processing_status: str = Field(serialization_alias="processingStatus")
    preview_available: bool = Field(serialization_alias="previewAvailable")


class QcDraftResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    version_id: UUID = Field(serialization_alias="id")
    version_number: int = Field(serialization_alias="versionNumber")
    title: str
    summary: str
    product_type: str = Field(serialization_alias="productType")
    content: str
    description: str
    manifest_hash: str = Field(serialization_alias="manifestHash")
    created_by_user_id: UUID = Field(serialization_alias="createdByUserId")
    created_at: datetime = Field(serialization_alias="createdAt")
    assets: list[QcDraftAssetResponse]


class QcDecisionResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    decision_id: UUID = Field(serialization_alias="id")
    status: str
    reason: str
    reviewer_user_id: UUID = Field(serialization_alias="reviewerUserId")
    checklist: list[QcChecklistItemResponse]
    created_at: datetime = Field(serialization_alias="createdAt")


class QcAgentCheckResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    key: str
    passed: bool
    detail: str


class QcAgentFindingResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    finding_id: UUID = Field(serialization_alias="id")
    category: str
    severity: str
    original_text: str = Field(serialization_alias="originalText")
    suggested_text: str = Field(serialization_alias="suggestedText")
    location: str
    detail: str
    confidence: float
    blocking: bool


class QcAgentPreflightResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    preflight_id: UUID = Field(serialization_alias="id")
    draft_version_id: UUID = Field(serialization_alias="draftVersionId")
    status: str
    checks: list[QcAgentCheckResponse]
    blockers: list[str]
    policy_version: str = Field(serialization_alias="policyVersion")
    created_at: datetime = Field(serialization_alias="createdAt")
    findings: list[QcAgentFindingResponse]


class QcIndexRecordResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    index_id: UUID = Field(serialization_alias="id")
    product_id: UUID = Field(serialization_alias="productId")
    status: str
    summary: str
    created_at: datetime = Field(serialization_alias="createdAt")


class QcDisseminationResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    dissemination_id: UUID = Field(serialization_alias="id")
    product_id: UUID = Field(serialization_alias="productId")
    recipient_user_id: UUID = Field(serialization_alias="recipientUserId")
    created_at: datetime = Field(serialization_alias="createdAt")


class QcFeedbackRequestResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    request_id: UUID = Field(serialization_alias="id")
    product_id: UUID = Field(serialization_alias="productId")
    requester_user_id: UUID = Field(serialization_alias="requesterUserId")
    status: str
    created_at: datetime = Field(serialization_alias="createdAt")


class QcProductSummaryResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    product_id: UUID = Field(serialization_alias="id")
    reference: str
    title: str
    status: str
    acg_ids: list[UUID] = Field(serialization_alias="acgIds")


class QcProductResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    ticket_id: UUID = Field(serialization_alias="ticketId")
    reference: str
    requester_user_id: UUID = Field(serialization_alias="requesterUserId")
    state: str
    title: str
    operational_question: str | None = Field(serialization_alias="operationalQuestion")
    area_or_region: str | None = Field(serialization_alias="areaOrRegion")
    priority: str | None
    priority_assessment: PriorityAssessmentResponse = Field(
        serialization_alias="priorityAssessment"
    )
    required_output_format: str | None = Field(serialization_alias="requiredOutputFormat")
    checklist_keys: list[str] = Field(serialization_alias="checklistKeys")
    latest_draft: QcDraftResponse | None = Field(serialization_alias="latestDraft")
    manager_notes: list[str] = Field(serialization_alias="managerNotes")
    decisions: list[QcDecisionResponse]
    agent_preflight: QcAgentPreflightResponse | None = Field(serialization_alias="agentPreflight")
    index_records: list[QcIndexRecordResponse] = Field(serialization_alias="indexRecords")
    disseminations: list[QcDisseminationResponse]
    feedback_requests: list[QcFeedbackRequestResponse] = Field(
        serialization_alias="feedbackRequests"
    )
    ingested_product: QcProductSummaryResponse | None = Field(serialization_alias="ingestedProduct")


class QcQueueItemResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    ticket_id: UUID = Field(serialization_alias="ticketId")
    reference: str
    state: str
    claim_status: str = Field(serialization_alias="claimStatus")


class QcQueueResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    products: list[QcProductResponse]
    items: list[QcQueueItemResponse] = Field(default_factory=list)
